### PR TITLE
tenancy: fully qualify type references

### PR DESCRIPTION
**Pull request by stevekuznetsov**
config: generate APIResourceSchemas and APIExports robustly

The generator is now reentrant and idempotent, and handles pruning.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

tenancy: all of clusterworkspacetype.spec is immutable

Initializers and type extensions were already immutable. We need
child/parent/default type references to be, as well. Additional labels
should be removed soon. Easier to make entirety of spec immutable rather
than allowing labels to be changed.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

tenancy: fully-qualify type relationships

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

tenancy: roll up conditions into ready

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

clusterworkspacetype: be explicitly permissive

Now that we have fully qualified references, finding some magic value
that is sensible for users and passes validation is getting silly. Just
use a boolean field.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

clusterwokspacetype: make default nullable

Otherwise you can't omit it and the field is not optional.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---

Part of #1149 
Depends on #1422 
/assign @sttts 
/cc @ncdc @davidfestal 

Follow-ups:

- [ ] permissive defaulting @sttts 
- [ ] union type check for any vs. types @sttts 
- [ ] CWT controller must ensure that allowed children and parent sets are subsets of extensions, if present
- [ ] nested loop over aliases when checking allowed children/parents @stevekuznetsov 
- [ ] fix requeuing and recomputation @stevekuznetsov

### CODE REVIEW

File: config/crds/tenancy.kcp.dev_clusterworkspacetypes.yaml [hunk 1]
```diff
@@ -53,37 +53,83 @@ spec:
                 description: additionalWorkspaceLabels are a set of labels that will
                   be added to a ClusterWorkspace on creation.
                 type: object
+              allowAnyChildWorkspaceTypes:
+                description: allowAnyChildWorkspaceTypes permits all ClusterWorkspaceTypes
+                  to be created in a workspace of this type.
+                type: boolean
+              allowAnyParentWorkspaceTypes:
```
Comment (sttts) [inline config/crds/tenancy.kcp.dev_clusterworkspacetypes.yaml:60]: I like this. It allows us sensible defaults (through admission 🤷‍♂️). Just a sum type would make this nicer:
```
allowedChildren:
  any: {}
  types: [{path: ..., name: ...}]
```

**Comment by stevekuznetsov** [comment]
@sttts updated to use a sum type

File: pkg/apis/tenancy/v1alpha1/types.go [hunk 2]
```diff
@@ -546,3 +549,15 @@ const (
...
+
+var (
+	// AnyWorkspaceTypeReference is used in allowed child and parent type lists to denote that any type is permissible.
+	AnyWorkspaceTypeReference = ClusterWorkspaceTypeReference{
+		Name: AnyWorkspaceType,
+	}
+	// RootWorkspaceTypeReference is a reference to the root logical cluster, which has no cluster workspace type
+	RootWorkspaceTypeReference = ClusterWorkspaceTypeReference{
+		Name: RootWorkspaceType,
+		Path: RootCluster.String(),
```
Comment (sttts) [inline pkg/apis/tenancy/v1alpha1/types.go:578]: nit: technically would have expected empty string here as another special case

File: pkg/reconciler/tenancy/clusterworkspacetype/clusterworkspacetype_controller_reconcile.go [hunk 3]
```diff
@@ -411,3 +428,48 @@ func (s referenceSet) UnsortedList() []tenancyv1alpha1.ClusterWorkspaceTypeRefer
 	}
 	return res
 }
+
+// resolveTypeRelationships ensures that:
+// - the default child type and allowed child types allow this owning type as a parent
+//   - the default child type is an allowed child type
+// - the allowed parent types allow this owning type as a child
+func (c *controller) resolveTypeRelationships(cwt *tenancyv1alpha1.ClusterWorkspaceType) error {
```
Comment (sttts) [inline pkg/reconciler/tenancy/clusterworkspacetype/clusterworkspacetype_controller_reconcile.go:436]: nit: resolve is a strange word for what this func is doing. It is rather validating than resolving. Resolve would either return something (non-error) or mutate.

File: pkg/apis/tenancy/v1alpha1/types.go [hunk 4]
```diff
@@ -204,23 +204,34 @@ type ClusterWorkspaceTypeSpec struct {
...
+	// permissive with child types.
 	//
 	// +optional
 	AllowedChildWorkspaceTypes []ClusterWorkspaceTypeReference `json:"allowedChildWorkspaceTypes,omitempty"`
 
+	// allowAnyParentWorkspaceTypes permits this type to be created in workspaces
+	// of all other ClusterWorkspaceTypes.
+	//
+	// +optional
+	AllowAnyParentWorkspaceTypes bool `json:"allowAnyParentWorkspaceTypes,omitempty"`
```
Comment (sttts) [inline pkg/apis/tenancy/v1alpha1/types.go]: my comment still stands: the intention is good, but this API now stutters. Make it
```
allowedParents:
  anyType: true # either this
  types: [...] # or this
```

File: config/root/clusterworkspacetype-team.yaml [hunk 5]
```diff
@@ -7,7 +7,9 @@ spec:
   defaultChildWorkspaceType:
     name: Universal
     path: root
-  allowAnyChildWorkspaceTypes: true
-  allowedParentWorkspaceTypes:
+  allowedChildren:
+    any: true
```
Comment (sttts) [inline config/root/clusterworkspacetype-team.yaml:11]: isn't (can't) this be defaulted?

File: pkg/admission/clusterworkspacetypeexists/admission.go [hunk 6]
```diff
@@ -196,17 +196,23 @@ func (o *clusterWorkspaceTypeExists) resolveValidType(parentClusterName logicalc
...
-	if !cwt.Spec.AllowAnyParentWorkspaceTypes && !setContainsAnyType(cwt.Spec.AllowedParentWorkspaceTypes, parentCwt.Status.TypeAliases) {
+	if !selectorMatchesAnyType(cwt.Spec.AllowedParents, parentCwt.Status.TypeAliases) {
 		return nil, fmt.Errorf("cluster workspace %q (of type %s) does not allow for parent workspaces of type %s", workspaceName, ref.String(), parentRef.String())
 	}
 	return cwt, nil
 }
 
-func setContainsAnyType(set, queries []tenancyv1alpha1.ClusterWorkspaceTypeReference) bool {
-	for _, allowed := range set {
+func selectorMatchesAnyType(selector *tenancyv1alpha1.ClusterWorkspaceTypeSelector, queries []tenancyv1alpha1.ClusterWorkspaceTypeReference) bool {
```
Comment (sttts) [inline pkg/admission/clusterworkspacetypeexists/admission.go:208]: ignore my comment in the other commit that this surprised. Better now.

File: pkg/apis/tenancy/v1alpha1/types.go [hunk 7]
```diff
@@ -205,27 +205,25 @@ type ClusterWorkspaceTypeSpec struct {
...
+	// workspace of this type.
 	//
 	// By default, no type is allowed. This means no other workspace can be nested
-	// within a workspace of the given type. Use allowAnyChildWorkspaceTypes to be
+	// within a workspace of the given type. Use allowedChildren.any=true to be
 	// permissive with child types.
 	//
 	// +optional
-	AllowedChildWorkspaceTypes []ClusterWorkspaceTypeReference `json:"allowedChildWorkspaceTypes,omitempty"`
+	AllowedChildren *ClusterWorkspaceTypeSelector `json:"allowedChildren,omitempty"`
```
Comment (sttts) [inline pkg/apis/tenancy/v1alpha1/types.go:216]: can we switch the default? We should be permissive in my eyes, and you nice new types allow that. OpenAPI can do.

File: config/crds/tenancy.kcp.dev_clusterworkspacetypes.yaml-patch [hunk 8]
```diff
@@ -11,6 +11,3 @@
         - root
         - system
         - any
-- op: add
-  path: /spec/versions/name=v1alpha1/schema/openAPIV3Schema/properties/spec/properties/allowedChildWorkspaceTypes/default
-  value: []
```
Comment (sttts) [inline config/crds/tenancy.kcp.dev_clusterworkspacetypes.yaml-patch:16]: add something to make any and types mutual exclusive.

File: pkg/apis/tenancy/v1alpha1/types.go (hunk 7 above)
Comment (sttts) [reply pkg/apis/tenancy/v1alpha1/types.go:216]: same below

File: config/crds/tenancy.kcp.dev_clusterworkspacetypes.yaml [hunk 9]
```diff
@@ -61,9 +61,22 @@ spec:
...
                 items:
-                  description: ClusterWorkspaceTypeName is a name of a ClusterWorkspaceType
-                  pattern: ^[A-Z][a-zA-Z0-9]+$
-                  type: string
+                  description: ClusterWorkspaceTypeReference is a globally unique,
+                    fully qualified reference to a cluster workspace type.
+                  properties:
+                    name:
+                      description: name is the name of the ClusterWorkspaceType
+                      pattern: ^[A-Z][a-zA-Z0-9]+$
```
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply config/crds/tenancy.kcp.dev_clusterworkspacetypes.yaml]: Let's make it orthogonal to this please, smaller blast radius.

File: pkg/admission/clusterworkspacetypeexists/admission.go [hunk 10]
```diff
@@ -197,23 +197,16 @@ func (o *clusterWorkspaceTypeExists) resolveValidType(parentClusterName logicalc
...
 		return nil, fmt.Errorf("parent cluster workspace %q (of type %s) does not allow for child workspaces of type %s", parentClusterName, parentRef.String(), ref.String())
 	}
-	if !setContainsAnyType(cwt.Spec.AllowedParentWorkspaceTypes, parentCwt.Status.TypeAliases) {
+	if !cwt.Spec.AllowAnyParentWorkspaceTypes && !setContainsAnyType(cwt.Spec.AllowedParentWorkspaceTypes, parentCwt.Status.TypeAliases) {
 		return nil, fmt.Errorf("cluster workspace %q (of type %s) does not allow for parent workspaces of type %s", workspaceName, ref.String(), parentRef.String())
 	}
 	return cwt, nil
 }
 
 func setContainsAnyType(set, queries []tenancyv1alpha1.ClusterWorkspaceTypeReference) bool {
```
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply pkg/admission/clusterworkspacetypeexists/admission.go]: Similar to `sets.String().HasAny()` - how would you prefer this to be clarified?

File: config/root/clusterworkspacetype-team.yaml (hunk 5 above)
Comment (stevekuznetsov) [reply config/root/clusterworkspacetype-team.yaml:11]: Isn't it? Defaults to false, keeps the behavior we had in the past.

**Pull request merged**